File: Antidote/ProfileManager.m
//
//  ProfileManager.m
//  Antidote
//
//  Created by Dmytro Vorobiov on 07.06.15.
//  Copyright (c) 2015 dvor. All rights reserved.
//

#import <objcTox/OCTDefaultFileStorage.h>
#import <objcTox/OCTManagerConfiguration.h>

#import "ProfileManager.h"
#import "NSArray+BlocksKit.h"
#import "UserDefaultsManager.h"

static NSString *const kSaveDirectoryPath = @"saves";

@interface ProfileManager ()

@property (strong, nonatomic, readwrite) NSArray *allProfiles;

@end

@implementation ProfileManager

#pragma mark -  Lifecycle

- (instancetype)init
{
    self = [super init];

    if (! self) {
        return nil;
    }

    [self reloadAllProfiles];

    return self;
}

#pragma mark -  Methods

- (BOOL)createProfileWithName:(NSString *)name error:(NSError **)error
{
    NSAssert(name.length > 0, @"name cannot be empty");

    NSString *path = [[self saveDirectoryPath] stringByAppendingPathComponent:name];

    NSFileManager *fileManager = [NSFileManager defaultManager];

    if (! [fileManager createDirectoryAtPath:path withIntermediateDirectories:YES attributes:nil error:error]) {
        return NO;
    }

    [self reloadAllProfiles];

    return YES;
}

- (BOOL)deleteProfileWithName:(NSString *)name error:(NSError **)error
{
    NSAssert(name.length > 0, @"name cannot be empty");

    NSString *path = [[self saveDirectoryPath] stringByAppendingPathComponent:name];

    NSFileManager *fileManager = [NSFileManager defaultManager];
    if (! [fileManager removeItemAtPath:path error:error]) {
        return NO;
    }

    [self reloadAllProfiles];

    return YES;
}

- (BOOL)renameProfileWithName:(NSString *)name toName:(NSString *)toName error:(NSError **)error
{
    NSAssert(name.length > 0, @"name cannot be empty");
    NSAssert(toName.length > 0, @"toName cannot be empty");

    NSString *fromPath = [[self saveDirectoryPath] stringByAppendingPathComponent:name];
    NSString *toPath = [[self saveDirectoryPath] stringByAppendingPathComponent:toName];

    if (! [[NSFileManager defaultManager] moveItemAtPath:fromPath toPath:toPath error:error]) {
        return NO;
    }

    [self reloadAllProfiles];

    return YES;
}

- (OCTManagerConfiguration *)configurationForProfileWithName:(NSString *)name passphrase:(NSString *)passphrase
{
    NSString *path = [[self saveDirectoryPath] stringByAppendingPathComponent:name];

    BOOL isDirectory = NO;
    BOOL exists = [[NSFileManager defaultManager] fileExistsAtPath:path isDirectory:&isDirectory];

    if (! exists || ! isDirectory) {
        return nil;
    }

    OCTManagerConfiguration *configuration = [OCTManagerConfiguration defaultConfiguration];
    configuration.passphrase = passphrase;

    configuration.options.IPv6Enabled = [AppContext sharedContext].userDefaults.uIpv6Enabled.boolValue;
    configuration.options.UDPEnabled = [AppContext sharedContext].userDefaults.uUDPEnabled.boolValue;

    configuration.fileStorage = [[OCTDefaultFileStorage alloc] initWithBaseDirectory:path
                                                                  temporaryDirectory:NSTemporaryDirectory()];

    return configuration;
}

#pragma mark -  Private

- (NSString *)saveDirectoryPath
{
    NSString *path = [NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) firstObject];
    return [path stringByAppendingPathComponent:kSaveDirectoryPath];
}

- (void)reloadAllProfiles
{
    NSFileManager *fileManager = [NSFileManager defaultManager];
    NSString *savePath = [self saveDirectoryPath];
    NSArray *contents = [fileManager contentsOfDirectoryAtPath:savePath error:nil];

    self.allProfiles = [contents bk_select:^BOOL (NSString *name) {
        NSString *path = [savePath stringByAppendingPathComponent:name];
        BOOL isDirectory;

        [fileManager fileExistsAtPath:path isDirectory:&isDirectory];

        return isDirectory;
    }];
}

@end

File: Antidote/ProfileManager.swift
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

import Foundation

private struct Constants {
    static let SaveDirectoryPath = "saves"

    // TODO get this constant from objcTox OCTDefaultFileStorage
    static let ToxFileName = "save.tox"
}

class ProfileManager {
    fileprivate(set) var allProfileNames: [String]

    init() {
        allProfileNames = []

        reloadProfileNames()
    }

    func createProfileWithName(_ name: String, copyFromURL: URL? = nil) throws {
        let path = pathFromName(name)
        let fileManager = FileManager.default

        try fileManager.createDirectory(atPath: path, withIntermediateDirectories: true, attributes: nil)

        if let url = copyFromURL {
            let saveURL = URL(fileURLWithPath: path).appendingPathComponent(Constants.ToxFileName)
            try fileManager.moveItem(at: url, to: saveURL)
        }

        reloadProfileNames()
    }

    func deleteProfileWithName(_ name: String) throws {
        let path = pathFromName(name)

        try FileManager.default.removeItem(atPath: path)

        reloadProfileNames()
    }

    func renameProfileWithName(_ fromName: String, toName: String) throws {
        let fromPath = pathFromName(fromName)
        let toPath = pathFromName(toName)

        try FileManager.default.moveItem(atPath: fromPath, toPath: toPath)

        reloadProfileNames()
    }

    func pathForProfileWithName(_ name: String) -> String {
        return pathFromName(name)
    }
}

private extension ProfileManager {
    func saveDirectoryPath() -> String {
        let path: NSString = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true).first! as NSString
        return path.appendingPathComponent(Constants.SaveDirectoryPath)
    }

    func pathFromName(_ name: String) -> String {
        let directoryPath: NSString = saveDirectoryPath() as NSString
        return directoryPath.appendingPathComponent(name)
    }

    func reloadProfileNames() {
        let fileManager = FileManager.default
        let savePath = saveDirectoryPath()

        let contents = try? fileManager.contentsOfDirectory(atPath: savePath)

        allProfileNames = contents?.filter {
            let path = (savePath as NSString).appendingPathComponent($0)

            var isDirectory: ObjCBool = false
            fileManager.fileExists(atPath: path, isDirectory:&isDirectory)

            return isDirectory.boolValue
        } ?? [String]()
    }
}
